### DISCUSSION TITLE
feat: Add Swagger for building REST API documentation

### DISCUSSION THREAD

**Discussion by ittuann** [apache/streampipes#2923] [Ideas]
In a recent discussion <https://github.com/apache/streampipes/discussions/2613>, I found that one of the needs is looking for tools or methods that can be used across multiple programming languages (Java, Python, Go). I considered using Swagger to build REST API documentation.

[Swagger](https://swagger.io/) (OpenAPI) is a language-agnostic specification, which means it is not limited to any specific programming language.

Swagger is used to describe REST APIs and provides a standardized way to detail an API, including paths, operations, input parameters, output responses, and more.

Not only does Swagger facilitate developers' understanding of the API's functionality, it also allows for real-time testing on web pages. Official examples can be viewed here: Swagger PetStore https://petstore.swagger.io/

I think it's suitable for the current scene. I created a sample StreamPipes demo using Swagger to show the effect:

![image](https://github.com/apache/streampipes/assets/80267918/133eff85-f84c-42a3-83d3-a1778be95252)

![image](https://github.com/apache/streampipes/assets/80267918/e83077d3-4db0-40fc-8c3a-c7286bf1d2aa)

![gif](https://github.com/apache/streampipes/assets/80267918/e9a503a0-01e2-4702-9e1e-61a0254e4538)

The content in this demo is not yet comprehensive and may contain some errors; it is merely to showcase the general effect and potential value of building with Swagger.

Beyond the cross-language capabilities, I believe that introducing Swagger documentation could greatly facilitate user integration of StreamPipes in unsupported languages, such as Rust.

If this feature is approved, I would also be willing to commit the current results to the project.

**Comment by bossenti** [comment]
Hey @ittuann, 

Thank you very much for your proposal! 
We already have swagger documents available for StreamPipes. You can access them via the link at the bottom of the login screen. 

Frankly, not all of our REST endpoints are document properly yet, but the most important ones should be. In case you want to polish some endpoints to that extend, PRs are always welcome 🙂
